activesupport/lib/active_support/message_encryptor.rb

```diff
@@ -43,65 +50,181 @@ class InvalidMessage < StandardError; end
     # * <tt>:digest</tt> - String of digest to use for signing. Default is +SHA1+.
     # * <tt>:serializer</tt> - Object serializer to use. Default is +Marshal+.
     def initialize(secret, *signature_key_or_options)
-      options = signature_key_or_options.extract_options!
-      sign_secret = signature_key_or_options.first
+      @options = signature_key_or_options.extract_options!
       @secret = secret
-      @sign_secret = sign_secret
-      @cipher = options[:cipher] || 'aes-256-cbc'
-      @verifier = MessageVerifier.new(@sign_secret || @secret, digest: options[:digest] || 'SHA1', serializer: NullSerializer)
-      @serializer = options[:serializer] || Marshal
+      @sign_secret = signature_key_or_options.first || @secret
+      @alg = @options[:alg] || 'dir'
+      @enc = @options[:enc] || 'A128GCM'
+      @serializer = @options[:serializer] || Marshal
     end
 
     # Encrypt and sign a message. We need to sign the message in order to avoid
     # padding attacks. Reference: http://www.limited-entropy.com/padding-oracle-attacks.
-    def encrypt_and_sign(value)
-      verifier.generate(_encrypt(value))
+    def encrypt_and_sign(value, options = {})
+      @claims = Claims.new(payload: value, **options)
+      _encrypt(@claims.to_h)
     end
 
     # Decrypt and verify a message. We need to verify the message in order to
     # avoid padding attacks. Reference: http://www.limited-entropy.com/padding-oracle-attacks.
-    def decrypt_and_verify(value)
-      _decrypt(verifier.verify(value))
+    def decrypt_and_verify(message, options = {})
+      if message.include? '--'
+        legacy_encryptor = LegacyMessageEncryptor.new(@secret, @sign_secret, @options)
+        legacy_encryptor.decrypt_and_verify(message)
+      else
+        if message.count('.') == 4
+          claims = _decrypt(message).symbolize_keys
+          Claims.verify!(claims, options)
+        else
+          raise InvalidFormat, 'Invalid JWT Format.'
+        end
+      end
     end
 
     private
+      def _encrypt(value)
+        protected_header = @serializer.dump header
+        cipher.encrypt
+        cipher.key = content_encryption_key
+        iv = cipher.random_iv
+        cipher.iv = iv
+        cipher.auth_data = to_ascii encode(protected_header)
+
+        ciphertext = cipher.update(@serializer.dump value) + cipher.final
+        auth_tag = cipher.auth_tag
+
+        ([protected_header, jwe_encrypted_key, iv, ciphertext, auth_tag].map { |a| encode a }).join('.')
```

Comment:
Don't need the parens around the array + map. map returns the array, you're joining